File: Difficulty: Medium/Spirally traversing a matrix/spirally-traversing-a-matrix.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    vector<int> spirallyTraverse(vector<vector<int> > &matrix) {
        vector<int> result;
        if(matrix.empty()) return result;
        
        int rows = matrix.size();
        int cols = matrix[0].size();
        
        int top = 0, bottom = rows - 1;
        int left = 0, right = cols - 1;
        
        while(top <= bottom && left <= right) {
            // Traverse from left to right on the topmost row
            for(int i = left; i <= right; i++) {
                result.push_back(matrix[top][i]);
            }
            top++;
            
            // Traverse from top to bottom on the rightmost column
            for(int i = top; i <= bottom; i++) {
                result.push_back(matrix[i][right]);
            }
            right--;
            
            // Make sure we are now on a different row
            if(top <= bottom) {
                // Traverse from right to left on the bottommost row
                for(int i = right; i >= left; i--) {
                    result.push_back(matrix[bottom][i]);
                }
                bottom--;
            }
            
            // Make sure we are now on a different column
            if(left <= right) {
                // Traverse from bottom to top on the leftmost column
                for(int i = bottom; i >= top; i--) {
                    result.push_back(matrix[i][left]);
                }
                left++;
            }
        }
        
        return result;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;

    while (t--) {
        int r, c;
        cin >> r >> c;
        vector<vector<int>> matrix(r, vector<int>(c, 0));

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                cin >> matrix[i][j];
            }
        }

        Solution ob;
        vector<int> result = ob.spirallyTraverse(matrix);
        for (int i = 0; i < result.size(); ++i)
            cout << result[i] << " ";
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Spirally traversing a matrix/spirally-traversing-a-matrix.py
class Solution:
    # Function to return a list of integers denoting spiral traversal of matrix.
    def spirallyTraverse(self, mat):
        # code here
        m, n = len(mat), len(mat[0])

    # List to store the spiral order elements
        res = []
    
        # Initialize boundaries
        top, bottom, left, right = 0, m - 1, 0, n - 1
    
        # Iterate until all elements are printed
        while top <= bottom and left <= right:
    
            # Print top row from left to right
            for i in range(left, right + 1):
                res.append(mat[top][i])
            top += 1
    
            # Print right column from top to bottom
            for i in range(top, bottom + 1):
                res.append(mat[i][right])
            right -= 1
    
            # Print bottom row from right to left (if exists)
            if top <= bottom:
                for i in range(right, left - 1, -1):
                    res.append(mat[bottom][i])
                bottom -= 1
    
            # Print left column from bottom to top (if exists)
            if left <= right:
                for i in range(bottom, top - 1, -1):
                    res.append(mat[i][left])
                left += 1
    
        return res

#{ 
 # Driver Code Starts
if __name__ == "__main__":
    import sys
    input = sys.stdin.read
    data = input().split()

    t = int(data[0])
    index = 1
    for _ in range(t):
        r = int(data[index])
        c = int(data[index + 1])
        index += 2
        matrix = []
        for i in range(r):
            row = list(map(int, data[index:index + c]))
            matrix.append(row)
            index += c

        solution = Solution()
        result = solution.spirallyTraverse(matrix)
        print(" ".join(map(str, result)))
        print("~")

# } Driver Code Ends